File: models/house/test_browser_materials.mjs
import assert from 'node:assert/strict';
import * as THREE from '../../house-test/vendor/three.module.min.js';
import {createHouseMaterial,finishDescription} from '../../house-test/materials.mjs';
import {segmentBlocked,choosePracticalLights} from '../../house-test/lighting.mjs';

const group=(name,finish)=>({name:'Fixture / '+name,color:[.25,.5,.75],glass:name.endsWith('glass'),finish});
const screen=createHouseMaterial(group('Dark appliance glass'));
assert.equal(screen.transparent,false);assert.equal(screen.opacity,1);
assert.equal(screen.depthWrite,true);assert(screen.roughness<.15);
const window=createHouseMaterial(group('Window glass'));
assert.equal(window.transparent,true);assert(window.opacity<.4);
assert.equal(window.depthWrite,false);
const steel=createHouseMaterial(group('Brushed stainless'));
assert.equal(steel.metalness,1);assert.equal(steel.color.r,.25,'Linear RGB was converted twice');
const lamp=createHouseMaterial(group('Warm glowing bulb',{surface:'plain',roughness:.4,
  emissive:[1,.79,.52],emissiveIntensity:3}));
assert.equal(lamp.emissive.r,1);assert.equal(lamp.emissiveIntensity,3);
assert.equal(finishDescription(group('New finish',{surface:'fabric',roughness:.94})).roughness,.94);
const panels=createHouseMaterial(group('Weathered dark exterior panels',{surface:'panels',roughness:.82,
  panelSize:[1.2,.6,.01],panelOffset:0,mortarColor:[.48,.44,.36]}));
const panelShader={vertexShader:THREE.ShaderLib.physical.vertexShader,
  fragmentShader:THREE.ShaderLib.physical.fragmentShader,defines:{},uniforms:{}};
panels.onBeforeCompile(panelShader);
assert.equal(panelShader.uniforms.houseSurfaceKind.value,13);
assert.deepEqual(panelShader.uniforms.housePanelSize.value.toArray(),[1.2,.6,.01]);
assert.equal(panelShader.uniforms.housePanelOffset.value,0);
assert.equal(panelShader.uniforms.houseMortarColor.value.r,.48);
assert.equal(panelShader.uniforms.houseDetailMap.value.generateMipmaps,true);
assert.equal(panelShader.uniforms.houseDetailMap.value.image.width,256);
const samePanels=createHouseMaterial(group('Second panel wall',{surface:'panels',panelSize:[1.2,.6,.01]}));
const sameShader={vertexShader:THREE.ShaderLib.physical.vertexShader,
  fragmentShader:THREE.ShaderLib.physical.fragmentShader,uniforms:{}};
samePanels.onBeforeCompile(sameShader);
assert.equal(sameShader.uniforms.houseDetailMap.value,panelShader.uniforms.houseDetailMap.value,
  'Repeated room groups must share finish texture memory');
samePanels.dispose();
panels.dispose();

// Long wood fibres should survive tiling without visible edge lines, and
// their subtle modulation must preserve each exported board/cabinet colour.
const wood=createHouseMaterial(group('Oak floor board tone 02',{surface:'wood'}));
const woodShader={vertexShader:THREE.ShaderLib.physical.vertexShader,
  fragmentShader:THREE.ShaderLib.physical.fragmentShader,uniforms:{}};
wood.onBeforeCompile(woodShader);
const {data:woodPixels,width:woodSize}=woodShader.uniforms.houseDetailMap.value.image;
const tone=(x,y)=>woodPixels[(y*woodSize+x)*4];
let along=0,across=0,edgeX=0,edgeY=0;
for(let y=0;y<woodSize;y++)for(let x=0;x<woodSize;x++){
  const value=tone(x,y)*2/255;
  assert(value>.85&&value<1.15,'Grain overwhelms the exported wood colour');
  if(x)along+=Math.abs(tone(x,y)-tone(x-1,y));
  if(y)across+=Math.abs(tone(x,y)-tone(x,y-1));
}
for(let i=0;i<woodSize;i++){
  edgeX+=Math.abs(tone(0,i)-tone(woodSize-1,i));
  edgeY+=Math.abs(tone(i,0)-tone(i,woodSize-1));
}
along/=woodSize*(woodSize-1);across/=woodSize*(woodSize-1);
assert(across>along*3,'Wood fibres lost their elongated direction');
assert(edgeX/woodSize<along*2+.5&&edgeY/woodSize<across*2+.5,
  'The repeating wood tile has a visible discontinuity at its boundary');
wood.dispose();

for(const surface of ['wood','fabric','carpet','blocks','siding','shakes','roof','lawn','mineral','stone','paint','ceramic','brushed','foliage']){
  const material=createHouseMaterial(group(surface,{surface,roughness:.6}));
  const shader={vertexShader:THREE.ShaderLib.physical.vertexShader,
    fragmentShader:THREE.ShaderLib.physical.fragmentShader,defines:{}};
  material.onBeforeCompile(shader);
  assert(shader.vertexShader.includes('vHousePosition=(modelMatrix*vec4(transformed,1.0)).xyz'));
  assert(shader.fragmentShader.includes('normal=houseBump(-vViewPosition,normal,houseDetail.z)'));
  assert(shader.uniforms.houseSurfaceKind.value>0);
  assert.equal(material.customProgramCacheKey(),panels.customProgramCacheKey(),
    'Finish families should share programs; the surface selector is a uniform');
  material.dispose();
}

const wall={min:[-.2,-1,1],max:[.2,4,1.2]};
assert(segmentBlocked([0,1,0],[0,2,3],[wall]));
assert(!segmentBlocked([1,1,0],[1,2,3],[wall]));
assert(!segmentBlocked([0,1,0],[0,1,.9],[wall]));
const source=(name,position)=>({name,position,power:40,type:'area'});
const lights=[source('Same room',[0,2,0]),source('Behind wall',[0,2,2]),
  source('Other floor',[0,5,0]),source('Distant',[10,2,0])];
assert.deepEqual(choosePracticalLights(lights,{x:0,y:0,z:0},[wall]).map(light=>light.name),['Same room']);
assert.equal(choosePracticalLights(Array.from({length:10},(_,i)=>source(String(i),[i*.1,2,0])),{x:0,y:0,z:0},[],3).length,3);
for(const material of [screen,window,steel,lamp])material.dispose();
console.log('PASS opaque displays, linear PBR finishes, all procedural shader hooks, room occlusion and bounded lights');

File: models/house/test_browser_materials.py
"""Finish/export contracts; no Blender runtime or private reference required."""
import unittest
from types import SimpleNamespace
from browser_materials import finish_for_name, material_finish, keep_bevel, practical_light, ramp_colliders


class BrowserMaterials(unittest.TestCase):
    def test_displays_are_opaque_and_windows_are_transparent(self):
        screen = finish_for_name('Dark appliance glass')
        window = finish_for_name('Window glass')
        self.assertEqual(screen['opacity'], 1)
        self.assertEqual(screen['surface'], 'screen')
        self.assertLess(window['opacity'], .4)
        self.assertEqual(window['surface'], 'glass')

    def test_finishes_have_meaningful_physical_differences(self):
        steel = finish_for_name('Brushed stainless')
        carpet = finish_for_name('Clean warm beige carpet')
        oak = finish_for_name('Oak floor board tone 02')
        self.assertEqual(steel['metalness'], 1)
        self.assertEqual(carpet['metalness'], 0)
        self.assertGreater(carpet['roughness'] - steel['roughness'], .5)
        self.assertGreater(oak['clearcoat'], .2)
        self.assertGreater(finish_for_name('Warm glowing bulb')['emissiveIntensity'], 1)

    def test_linked_shader_socket_does_not_export_its_unused_default(self):
        sockets = {'Roughness': SimpleNamespace(default_value=.5, is_linked=True),
                   'Metallic': SimpleNamespace(default_value=.8, is_linked=False)}
        node = SimpleNamespace(type='BSDF_PRINCIPLED', inputs=sockets)
        mat = SimpleNamespace(name='Brushed stainless', use_nodes=True,
                              node_tree=SimpleNamespace(nodes=[node]))
        finish = material_finish(mat)
        self.assertEqual(finish['roughness'], .3)
        self.assertEqual(finish['metalness'], .8)

    def test_architecture_does_not_spend_the_furniture_bevel_budget(self):
        self.assertTrue(keep_bevel('Drawer front', '08 | Kitchen cabinetry', (1, .04, .3), 8))
        for name, collection, size, vertices in [
            ('Oak floor board', '02 | Floors and split levels', (2, .1, .02), 8),
            ('Roof frame', '43 | Exterior roofs and cladding', (2, .2, .2), 8),
            ('Tiny trim', '08 | Kitchen cabinetry', (1, .008, .008), 8),
            ('Sofa cushion', '09 | Living furniture', (1, .5, .2), 400),
        ]:
            self.assertFalse(keep_bevel(name, collection, size, vertices))

    def test_panel_grid_comes_from_the_blender_finish_not_a_wood_assumption(self):
        values = {'Scale': 1, 'Brick Width': 1.2, 'Row Height': .6,
                  'Mortar Size': .010, 'Mortar': [.48, .44, .36, 1]}
        brick = SimpleNamespace(type='TEX_BRICK', offset=0,
            inputs={key: SimpleNamespace(default_value=value) for key, value in values.items()})
        mat = SimpleNamespace(name='Weathered dark exterior panels', use_nodes=True,
                              node_tree=SimpleNamespace(nodes=[brick]))
        finish = material_finish(mat)
        self.assertEqual(finish['surface'], 'panels')
        self.assertEqual(finish['panelSize'], [1.2, .6, .01])
        self.assertEqual(finish['panelOffset'], 0)
        self.assertEqual(finish['mortarColor'], [.48, .44, .36])

    def test_render_softboxes_are_not_house_lamps(self):
        self.assertFalse(practical_light('Front daylight', 'AREA', 1000))
        self.assertFalse(practical_light('Soft daylight sun', 'SUN', 2.2))
        self.assertTrue(practical_light('Living ceiling fill', 'AREA', 55))
        self.assertTrue(practical_light('Floor lamp warm bulb', 'POINT', 42))

    def test_apron_ramp_has_bounded_strips_in_browser_coordinates(self):
        points = [(-6.75,-4.8,-.795),(-.25,-4.8,-.795),
                  (-.25,-1.8,-.16),(-6.75,-1.8,-.16)]
        boxes = ramp_colliders('Apron', points, 'y')
        self.assertEqual(len(boxes), 30)
        self.assertAlmostEqual(boxes[0]['max'][2], 4.8)
        self.assertAlmostEqual(boxes[-1]['min'][2], 1.8)
        self.assertLess(abs(boxes[0]['max'][1] + .795), .012)
        self.assertLess(abs(boxes[-1]['max'][1] + .16), .012)
        for previous, box in zip(boxes, boxes[1:]):
            self.assertAlmostEqual(previous['min'][2], box['max'][2])
            self.assertLess(box['max'][1] - previous['max'][1], .03)

    def test_ramp_rejects_twisted_or_nonrectangular_meshes(self):
        with self.assertRaisesRegex(ValueError, 'planar'):
            ramp_colliders('Twisted', [(0,0,0),(1,0,0),(1,1,1),(0,1,2)], 'y')
        with self.assertRaisesRegex(ValueError, 'rectangular'):
            ramp_colliders('Triangle', [(0,0,0),(1,0,0),(.5,.5,.5),(0,1,1)], 'y')


if __name__ == '__main__':
    unittest.main()
